container-core/src/main/java/com/yahoo/container/di/Container.java

```diff
@@ -70,20 +70,26 @@ public Container(SubscriberFactory subscriberFactory, String configId, Component
     // TODO: try to simplify by returning the result even when the graph failed, instead of throwing here.
     public ComponentGraphResult waitForNextGraphGeneration(ComponentGraph oldGraph, Injector fallbackInjector, boolean isInitializing) {
         try {
+            ComponentGraph newGraph;
             Collection<Bundle> obsoleteBundles = new HashSet<>();
-            ComponentGraph newGraph = waitForNewConfigGenAndCreateGraph(oldGraph, fallbackInjector, isInitializing, obsoleteBundles);
-            newGraph.reuseNodes(oldGraph);
+            try {
+                newGraph = waitForNewConfigGenAndCreateGraph(oldGraph, fallbackInjector, isInitializing, obsoleteBundles);
+                newGraph.reuseNodes(oldGraph);
+            } catch (Throwable t) {
+                log.warning("Failed to set up component graph - uninstalling latest bundles. Bootstrap generation: " + getBootstrapGeneration());
+                osgi.completeBundleGeneration(Osgi.GenerationStatus.FAILURE);
+                throw t;
```

Comment:
Any started bundles here are never uninstalled, as in `bundle.uninstall()`, as far as I can see.

Reply:
It looks like it would be OK to create a similar failed graph deconstruction task.

Reply:
Correct. I'll see if I can fix it easily, or add a TODO. They are not installed today, and not in the previous attempted fix either.